### PR TITLE
src/components/routes: add component RoutesCalendar to display routes to be logged by users

### PR DESCRIPTION
**Pull request by maceksimon**
Install the [QCalendar](https://qcalendar.netlify.app/) extension by using the [app extension way](https://qcalendar.netlify.app/all-about-qcalendar/installation-types#Quasar-CLI), for simplicity.
(In component tests, QCalendarMonth has to be included manually as a Vue plugin.)

Use the `QCalendarMonth` component to render the calendar UI.
Use the day slot to render routes to work and from work for each past day.
Add a `CalendarNavigation` component to display top navigation buttons using native QCalendar methods.
Add a `CalendarItemDisplay` component to display each individual route (to work or from work) based on states (active/not-active, logged/not-logged). Makes use of custom SVGs for background.
Add calendar fixture.
Add tests.

Screenshot 2024-06-30 at 14 11 44

### CODE REVIEW

File: src/components/routes/CalendarItemDisplay.vue [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+      <!-- Background: To work -->
+      <template v-if="direction === 'toWork'">
+        <!-- Icon: To work - active -->
+        <q-icon
+          v-if="active"
+          color="primary"
+          class="full-width full-height absolute-full"
+          data-cy="calendar-item-icon-towork-active"
+        >
+          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 135 38">
```
Comment (tmszi) [inline src/components/routes/CalendarItemDisplay.vue:113]: Should be standalone SVG image.

File: src/components/routes/CalendarItemDisplay.vue [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        </q-icon>
+        <!-- Icon: To work - logged -->
+        <q-icon
+          v-else-if="route && route.id"
+          color="secondary"
+          class="full-width full-height absolute-full"
+          style="opacity: 0.4"
+          data-cy="calendar-item-icon-towork-logged"
+        >
+          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 135 38">
```
Comment (tmszi) [inline src/components/routes/CalendarItemDisplay.vue:129]: Should be standalone SVG image.

File: src/components/routes/CalendarItemDisplay.vue [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        </q-icon>
+        <!-- Icon: To work - empty -->
+        <q-icon
+          v-else
+          color="primary"
+          class="full-width full-height absolute-full"
+          style="opacity: 0.5"
+          data-cy="calendar-item-icon-towork-empty"
+        >
+          <svg
```
Comment (tmszi) [inline src/components/routes/CalendarItemDisplay.vue:145]: Should be standalone SVG image.

File: src/components/routes/CalendarItemDisplay.vue [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+      <!-- Background: From work -->
+      <template v-if="direction === 'fromWork'">
+        <!-- Icon: From work - active -->
+        <q-icon
+          v-if="active"
+          color="primary"
+          class="full-width full-height absolute-full"
+          data-cy="calendar-item-icon-fromwork-active"
+        >
+          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 135 38">
```
Comment (tmszi) [inline src/components/routes/CalendarItemDisplay.vue:166]: Should be standalone SVG image.

File: src/components/routes/CalendarItemDisplay.vue [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        </q-icon>
+        <!-- Icon: From work - logged -->
+        <q-icon
+          v-else-if="route && route.id"
+          color="secondary"
+          class="full-width full-height absolute-full"
+          style="opacity: 0.4"
+          data-cy="calendar-item-icon-fromwork-logged"
+        >
+          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 135 38">
```
Comment (tmszi) [inline src/components/routes/CalendarItemDisplay.vue:182]: Should be standalone SVG image.

File: src/components/routes/CalendarItemDisplay.vue [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        </q-icon>
+        <!-- Icon: From work - empty -->
+        <q-icon
+          v-else
+          color="primary"
+          class="full-width full-height absolute-full"
+          style="opacity: 0.5"
+          data-cy="calendar-item-icon-fromwork-empty"
+        >
+          <svg
```
Comment (tmszi) [inline src/components/routes/CalendarItemDisplay.vue:198]: Should be standalone SVG image.

File: src/components/routes/CalendarItemDisplay.vue [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+        </q-icon>
+      </template>
+
+      <!-- Content: Route logged -->
+      <template v-if="route && route.id">
+        <!-- Icon: Transport type -->
+        <q-icon
+          :color="active ? 'white' : 'primary'"
+          :name="getRouteIcon(route.transport)"
+          size="18px"
```
Comment (tmszi) [inline src/components/routes/CalendarItemDisplay.vue:217]: size attr value 18px should be const (repeat > 1).

File: src/components/routes/CalendarItemDisplay.vue [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+          data-cy="calendar-item-icon-transport"
+        />
+        <!-- Distance -->
+        <span
+          class="relative-position text-subtitle2"
+          :class="[active ? 'text-white' : 'text-primary']"
+          data-cy="calendar-item-distance"
+        >
+          {{ route.distance }} km
```
Comment (tmszi) [inline src/components/routes/CalendarItemDisplay.vue:226]: Unit should be global var?

File: src/components/routes/CalendarNavigation.vue [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        outline
+        size="12px"
+        color="primary"
+        @click="onPrev"
+        data-cy="calendar-navigation-previous"
+      >
+        <q-icon
+          name="arrow_back"
+          color="primary"
+          size="18px"
```
Comment (tmszi) [inline src/components/routes/CalendarNavigation.vue:82]: size attr value 18px should be const (repeat > 1).

File: src/components/routes/RoutesCalendar.vue [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+    });
+
+    // Active state
+    const activeItem = ref<Timestamp | null>(parsed(today()));
+    const activeDirection = ref<TransportDirection>('toWork');
+
+    /**
+     * Determines if route item is active.
+     * It checks if the timestamp and direction against a stored state.
+     * @param param0 { timestamp: Timestamp; direction: TransportDirection }
```
Comment (tmszi) [inline src/components/routes/RoutesCalendar.vue:103]: `@param Object`

File: src/components/routes/RoutesCalendar.vue [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+        activeItem.value.date === timestamp.date &&
+        activeDirection.value === direction
+      );
+    }
+
+    /**
+     * Handles click on route item within a day frame.
+     * It triggers active state on that day.
+     * It controls content of the route-logging dialog panel.
+     * @param param0 { timestamp: Timestamp; direction: TransportDirection }
```
Comment (tmszi) [inline src/components/routes/RoutesCalendar.vue:130]: `@param Object`

File: src/components/routes/RoutesCalendar.vue [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+        return '';
+      }
+      const date = new Date(selectedDate.value);
+      const month = date.toLocaleString(locale.value, { month: 'long' });
+      const year = date.toLocaleString(locale.value, { year: 'numeric' });
+      return `${month} ${year}`;
+    });
+
+    // Calendar naviation functions
+    function onToday() {
```
Comment (tmszi) [inline src/components/routes/RoutesCalendar.vue:66]: missing return type void

File: src/components/routes/RoutesCalendar.vue [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+            />
+          </div>
+        </template>
+      </q-calendar-month>
+    </div>
+  </div>
+</template>
+
+<style lang="scss" scoped>
+:deep(.q-calendar-month__day--label) {
```
Comment (tmszi) [inline src/components/routes/RoutesCalendar.vue:229]: Really need this customization (potentially this code break up with next release)?

File: src/components/routes/CalendarItemDisplay.vue (hunk 1 above)
Comment (maceksimon) [reply src/components/routes/CalendarItemDisplay.vue:113]: Refactored using [public svg with svg:use](https://quasar.dev/vue-components/icon#svg-use-way)
Will be used for new custom SVGs.

File: src/components/routes/CalendarItemDisplay.vue (hunk 3 above)
Comment (maceksimon) [reply src/components/routes/CalendarItemDisplay.vue:145]: Refactored using [public svg with svg:use](https://quasar.dev/vue-components/icon#svg-use-way)
Will be used for new custom SVGs.

File: src/components/routes/CalendarItemDisplay.vue (hunk 5 above)
Comment (maceksimon) [reply src/components/routes/CalendarItemDisplay.vue:182]: Refactored using [public svg with svg:use](https://quasar.dev/vue-components/icon#svg-use-way)
Will be used for new custom SVGs.

File: src/components/routes/CalendarItemDisplay.vue (hunk 8 above)
Comment (maceksimon) [reply src/components/routes/CalendarItemDisplay.vue:226]: Yes, refactored into a new composable function in `useRoutes.ts`.

File: src/components/routes/CalendarItemDisplay.vue (hunk 2 above)
Comment (maceksimon) [reply src/components/routes/CalendarItemDisplay.vue:129]: Refactored using [public svg with svg:use](https://quasar.dev/vue-components/icon#svg-use-way)
Will be used for new custom SVGs.

File: src/components/routes/CalendarItemDisplay.vue (hunk 4 above)
Comment (maceksimon) [reply src/components/routes/CalendarItemDisplay.vue:166]: Refactored using [public svg with svg:use](https://quasar.dev/vue-components/icon#svg-use-way)
Will be used for new custom SVGs.

File: src/components/routes/CalendarItemDisplay.vue (hunk 6 above)
Comment (maceksimon) [reply src/components/routes/CalendarItemDisplay.vue:198]: Refactored using [public svg with svg:use](https://quasar.dev/vue-components/icon#svg-use-way)
Will be used for new custom SVGs.

File: src/components/routes/CalendarItemDisplay.vue (hunk 7 above)
Comment (maceksimon) [reply src/components/routes/CalendarItemDisplay.vue:217]: Updated.

File: src/components/routes/CalendarNavigation.vue (hunk 9 above)
Comment (maceksimon) [reply src/components/routes/CalendarNavigation.vue:82]: Updated.

File: src/components/routes/RoutesCalendar.vue (hunk 10 above)
Comment (maceksimon) [reply src/components/routes/RoutesCalendar.vue:103]: Fixed.

File: src/components/routes/RoutesCalendar.vue (hunk 11 above)
Comment (maceksimon) [reply src/components/routes/RoutesCalendar.vue:130]: Fixed.

File: src/components/routes/CalendarNavigation.vue [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+      data-cy="calendar-navigation-today"
+      :style="{ 'border-radius': borderRadius }"
+    />
+    <div class="flex gap-8">
+      <!-- Button: previous -->
+      <q-btn
+        unelevated
+        round
+        outline
+        size="12px"
```
[the remark this answers is not in the document]
Comment (maceksimon) [reply src/components/routes/CalendarNavigation.vue:74]: Updated.

File: src/components/routes/RoutesCalendar.vue (hunk 12 above)
Comment (maceksimon) [reply src/components/routes/RoutesCalendar.vue:66]: Added.

File: src/components/routes/RoutesCalendar.vue (hunk 13 above)
Comment (maceksimon) [reply src/components/routes/RoutesCalendar.vue:229]: I did not find a clean api for styling the highlighted (selected) date. Based on design it should have semi-transparent background color and no outline.
I will check again.
If there is no way of styling and we need to follow design closely, we can write tests for it (they should signal breaking change).

Comment (maceksimon) [reply src/components/routes/RoutesCalendar.vue]: This was resolved in the team discussion - we are moving forward with default styling.
I removed the styles.

File: src/components/routes/CalendarNavigation.vue [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+    <div class="flex gap-8">
+      <!-- Button: previous -->
+      <q-btn
+        unelevated
+        round
+        outline
+        :size="buttonSize"
+        color="primary"
+        @click="onPrev"
+        data-cy="calendar-navigation-previous"
```
Comment (tmszi) [inline src/components/routes/CalendarNavigation.vue:82]: `data-cy` should be `calendar-navidagation-previous-btn`

File: src/components/routes/CalendarNavigation.vue [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+      >
+        <q-icon
+          name="arrow_back"
+          color="primary"
+          :size="iconSize"
+          data-cy="calendar-navigation-prev"
```
Comment (tmszi) [inline src/components/routes/CalendarNavigation.vue:88]: `data-cy` should be `calendar-navidagation-previous-icon`

File: src/components/routes/CalendarNavigation.vue [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+      </q-btn>
+      <!-- Button: next -->
+      <q-btn
+        unelevated
+        round
+        outline
+        :size="buttonSize"
+        color="primary"
+        @click="onNext"
+        data-cy="calendar-navigation-next"
```
Comment (tmszi) [inline src/components/routes/CalendarNavigation.vue:99]: `data-cy` should be `calendar-navidagation-next-btn`

File: src/components/routes/RoutesCalendar.vue [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+
+    // Active state
+    const activeItem = ref<Timestamp | null>(parsed(today()));
+    const activeDirection = ref<TransportDirection>('toWork');
+
+    /**
+     * Determines if route item is active.
+     * It checks if the timestamp and direction against a stored state.
+     * @param Object { timestamp: Timestamp; direction: TransportDirection }
+     */
```
Comment (tmszi) [inline src/components/routes/RoutesCalendar.vue:106]: `* @return boolean`

File: src/composables/useRoutes.ts [hunk 19]
```diff
@@ -35,6 +35,11 @@ export const useRoutes = () => {
     }
   };
 
+  const getRouteDistance = (route: RouteItem | null): string => {
```
Comment (tmszi) [inline src/composables/useRoutes.ts:38]: Func doc

File: src/components/routes/CalendarNavigation.vue (hunk 15 above)
Comment (maceksimon) [reply src/components/routes/CalendarNavigation.vue:82]: Updated.

File: src/components/routes/CalendarNavigation.vue (hunk 16 above)
Comment (maceksimon) [reply src/components/routes/CalendarNavigation.vue:88]: Updated.

File: src/components/routes/CalendarNavigation.vue (hunk 17 above)
Comment (maceksimon) [reply src/components/routes/CalendarNavigation.vue:99]: Updated.

File: src/components/routes/CalendarNavigation.vue [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+      >
+        <q-icon
+          name="arrow_forward"
+          color="primary"
+          :size="iconSize"
+          data-cy="calendar-navigation-next"
```
[the remark this answers is not in the document]
Comment (maceksimon) [reply src/components/routes/CalendarNavigation.vue:105]: Updated.

File: src/composables/useRoutes.ts (hunk 19 above)
Comment (maceksimon) [reply src/composables/useRoutes.ts:44]: Added.

File: src/components/routes/RoutesCalendar.vue (hunk 18 above)
Comment (maceksimon) [reply src/components/routes/RoutesCalendar.vue:107]: Added.

**Pull request merged**